### PR TITLE
[Feature]: Add points structure

### PR DESCRIPTION
**Pull request by encore-zhou**
1. Add Classes `BaseInstancePoints`, `DepthInstancePoints`, `CameraInstancePoints`, `LiDARInstancePoints`, `Coord3DMode`.
2. Add unittest for each Class.

### CODE REVIEW

File: mmdet3d/core/points/__init__.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from .base_points import BaseInstancePoints
```
Comment (ZwwWayne) [inline mmdet3d/core/points/__init__.py:1]: BaseInstancePoints -> BasePoints
CameraInstancePoints -> CameraPoints
DepthInstancePoints -> DepthPoints
LiDARInstancePoints -> LiDARPoints
`Instance` here seems unnecessary.

File: mmdet3d/core/points/base_points.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
+import torch
+from abc import abstractmethod
+
+
+class BaseInstancePoints(object):
+    """Base class for Points.
+
+    Note:
+        The box is bottom centered, i.e. the relative position of origin in
```
Comment (ZwwWayne) [inline mmdet3d/core/points/base_points.py:9]: Update docstrings.

File: mmdet3d/core/points/base_points.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    def shuffle(self):
+        """Shuffle the points."""
+        self.tensor = self.tensor[torch.randperm(
+            self.__len__(), device=self.tensor.device)]
+
+    def rotate(self, rot_mat_T):
+        """Rotate points with the given rotation matrix.
+
+        Args:
+            rot_mat_T (np.ndarray, torch.Tensor): Rotation matrix.
```
Comment (ZwwWayne) [inline mmdet3d/core/points/base_points.py:82]: We may support either use `rot_mat_T` or a rotation angle for rotation.

**Comment by ZwwWayne** [comment]
According to the [coverage report](https://codecov.io/github/open-mmlab/mmdetection3d/commit/28b62ec0fab9751bc36c0b1c21e978716d48e185). The coverage of `coord3d_mode.py` is a little bit low.

File: mmdet3d/core/points/base_points.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+            2. `new_points = points[2:10]`:
+                return a slice of points.
+            3. `new_points = points[vector]`:
+                where vector is a torch.BoolTensor with `length = len(points)`.
+                Nonzero elements in the vector will be selected.
+            Note that the returned Points might share storage with this Points,
+            subject to Pytorch's indexing semantics.
+
+        Returns:
+            :obj:`BaseInstancesPints`: A new object of  \
```
Comment (ZwwWayne) [inline mmdet3d/core/points/base_points.py:211]: Update docstrings accordingly.

**Comment by hellock** [comment]
Task linked: [CU-4qth4e Add points structure](https://app.clickup.com/t/4qth4e)

File: mmdet3d/core/points/base_points.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
+import torch
+from abc import abstractmethod
+
+
+class BasePoints(object):
+    """Base class for Points.
+
+    Args:
+        tensor (torch.Tensor | np.ndarray | list): a N x points_dim matrix.
+        points_dim (int): Number of the dimension of a point.
```
Comment (ZwwWayne) [inline mmdet3d/core/points/base_points.py:10]: (int, optional)

File: mmdet3d/core/points/base_points.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+            Each row is (x, y, z). Default to 3.
+        attribute_dims (dict): Dictinory to indicate the meaning of extra
```
Comment (ZwwWayne) [inline mmdet3d/core/points/base_points.py:12]: (dict, optional)

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**